### PR TITLE
Use namespace, directives and attributes to compare equal of Requirement

### PR DESCRIPTION
**Pull request by laeubi**
Currently the resolver is ultra slow in some scenarios and even never finds a valid solution, just adding more and more permutations to check.

The reason is that one check that should actually *reduce* the number of problematic permutations fails because it uses Requirement#equals what is incapable of compare requirements across Resource boundaries, that means even if the requirement describes the same restrictions it is not considered equal and therefore selected as a possible permutation to remove this, then the validation check detects this is unresolvable adding possible new permutations and so on.

This now performs an equals check instead on namespace, directives and attributes (what can be disabled by a system property) what makes the resolver in such case from running forever to just an blink of an eye.

@tjwatson this _dramatically_ reduces resolving time and leads to a resolved state where previously confusing use constraint violation and missing packages are reported that are actually resolvable. For my example described here:

- https://github.com/eclipse-equinox/equinox/pull/469
- https://github.com/eclipse-equinox/equinox/issues/448

it goes from minutes (until it hits the timeout) to milliseconds, I therefore suggest to include this in RC1, I added a system property as last resort if this would break something badly, so please review and give +1 as PL, for the record I would also give +1 as PL but would want your vote and review as binding here!

FYI @tivervac

### CODE REVIEW

**Comment by laeubi** [comment]
In my example there is the following requirements from two bundles that are using substitution packages:

```
osgi.wiring.package; filter:="(osgi.wiring.package=org.eclipse.lsp4j.jsonrpc)"
osgi.wiring.package; filter:="(osgi.wiring.package=org.eclipse.lsp4j.jsonrpc)"
```

their toString looks equals but if you look into the object you will see that one requirement is from

`osgi.identity; osgi.identity="org.eclipse.lsp4j.jsonrpc"; type="osgi.bundle"; version:Version="0.19.0.v20221118-0359"`

and the other from

`osgi.identity; osgi.identity="org.eclipse.lsp4j.jsonrpc"; type="osgi.bundle"; version:Version="0.21.1.v20230829-0012"`

so using equals compare them as **false** resulting in (the drop of lines is where the resolver gives up after one minute):

grafik

after apply they fix this results in the following:

grafik

**Comment by stbischof** [comment]
@timothyjward fyi

File: bundles/org.eclipse.osgi/felix/src/org/apache/felix/resolver/Candidates.java [hunk 1]
```diff
@@ -1260,6 +1287,25 @@ public boolean canRemoveCandidate(Requirement req)
         return false;
     }
 
+	private boolean isMatchingRequirement(Requirement requirement, Requirement other) {
+		if (USE_REQUIREMENTS_EQUALS) {
```
Comment (stbischof) [inline bundles/org.eclipse.osgi/felix/src/org/apache/felix/resolver/Candidates.java:1291]: you may want to document this fallback somewhere

**Comment by tjwatson** [comment]
> so using equals compare them as **false** resulting in (the drop of lines is where the resolver gives up after one minute):

But they are two different requirements because they are from two different bundles.

**Comment by laeubi** [comment]
Convert to draft as it does not seem to solve the issue in all cases...

**Comment by laeubi** [comment]
@tjwatson I agree that this seems not the correct way to address the issue, I'll closing this for now and try to came up with an improved solution. This has most probably be detected in the inital collection of candidates so the resolver can't run into a conflict.

**Pull request closed without merging**